### ISSUE TITLE
BINTOPKGLIBEXECDIR - `tools/rel.sh` uses GNUisms in `sed` script

### ISSUE THREAD

**Issue by jsarenik** [ElementsProject/lightning#1221]
### Issue and Steps to Reproduce
This is a quick note that `make check` actually requires `make install` to run before it in order to do its work. Otherwise there is following error:

```
make check
...
make[2]: Leaving directory '/home/jasan/src/lightning'
testinstall/usr/local/bin/lightningd --test-daemons-only --lightning-dir=testinstall
lightningd: Could not run /usr/local/libexec/c-lightning/lightning_channeld: No such file or directory
make[1]: *** [Makefile:436: installcheck] Error 1
make[1]: Leaving directory '/home/jasan/src/lightning'
make: *** [Makefile:205: check] Error 2
```

Which IMHO points to file `lightningd/lightningd.c`, [line 183](https://github.com/ElementsProject/lightning/blob/master/lightningd/lightningd.c#L183) - the line with `BINTOPKGLIBEXECDIR`.

**Comment by ZmnSCPxj** [comment]
That is very strange, since it passes in CI, which does not perform install before check.

Note that check itself performs an install check where it checks that even if installed in a nonstandard location, it can still run. Cam you provide log of `make check` run?

**Comment by jsarenik** [comment]
I see your point and now I am investigating further. Will be back with full log if it makes sense, but so far I have two different results on two machines. Tracking down the differences between them...

**Comment by ZmnSCPxj** [comment]
When compiling, what is the `-DBINTOPKGLIBEXECDIR=` value?  Also, can you execute this command in the lightningd tree?

    tools/rel.sh /usr/local/bin /usr/local/libexec/c-lightning

**Comment by jsarenik** [comment]
Sorry, not answering your question as I want to show one intermediate result now:

The compilation procudure is completely the same on both systems. Both systems are musl-based and up-to-date Alpine Linuxes.

On the system where it is failing:

```
$ strings lightningd/lightningd | grep libexec
/../../..//usr/local/libexec/c-lightning
lightningd/lightningd.c:184:pkglibexecdir
pkglibexecdir
find_my_pkglibexec_path
find_my_pkglibexec_path
```

On the system which passes `make check`:

```
$ strings lightningd/lightningd | grep libexec
../libexec/c-lightning
lightningd/lightningd.c:184:pkglibexecdir
pkglibexecdir
find_my_pkglibexec_path
find_my_pkglibexec_path
```

**Comment by jsarenik** [comment]
The output of the command you wrote @ZmnSCPxj is exactly as the first line in my `string` outputs above.

**Comment by jsarenik** [comment]
It is because on one system I have GNU Sed and on another just Busybox sed.

**Comment by jsarenik** [comment]
TLDR install GNU Sed.

Another possibility would be to simplify the sed string so that it works on Busybox sed as well. Let me see...

**Comment by ZmnSCPxj** [comment]
Looks like I accidentally used a GNUism in `tools/rel.sh`.  I shall have to look into this more.

**Comment by ZmnSCPxj** [comment]
Can you try replacing `tools/rel.sh` with this on the Busybox `sed`?

    #! /bin/sh
    
    from=${1}
    to=${2}
    common=`printf '%s\0%s' "${from}" "${to}" | sed 's/\(.*\).*\x0\1.*$/\1/' | sed 's@/[^/]*$@/@'`
    prefix=`printf '%s\n' ${from#$common} | sed 's@[^/][^/]*@..@g'`
    printf '%s\n' "$prefix/${to#$common}"

**Comment by jsarenik** [comment]
@ZmnSCPxj doesn't help:

```
$ sh -x tools/rel.sh /usr/local/bin /usr/local/libexec/c-l
ightning
+ from=/usr/local/bin
+ to=/usr/local/libexec/c-lightning
+ printf '%s\0%s' /usr/local/bin /usr/local/libexec/c-lightning
+ sed 's/\(.*\).*\x0\1.*$/\1/'
+ sed 's@/[^/]*$@/@'
+ common=/usr/local//usr/local/libexec/
+ printf '%s\n' /usr/local/bin
+ sed 's@[^/][^/]*@..@g'
+ prefix=/../../..
+ printf '%s\n' /../../..//usr/local/libexec/c-lightning
/../../..//usr/local/libexec/c-lightning
```

**Comment by jsarenik** [comment]
By the way what should be the result of `(.*).*`?

**Comment by jsarenik** [comment]
Feel free to use statically linked binaries from https://busybox.net/downloads/binaries/1.28.1-defconfig-multiarch/ as I need to go offline now.

**Comment by jsarenik** [comment]
@ZmnSCPxj and thank you! It is really not essential right now for me, but I appreciate your interest in the issue. I mostly wanted to figure out what was the reason it behaved differently.

**Comment by ZmnSCPxj** [comment]
> By the way what should be the result of `\(.*\).*`?

`.*` is greedy and tries to consume everything, so normally `\(.*\)` will match everything and leave nothing for `.*` to match.  However, we have, a `\1` after the separator character (separator was `\0` originally and matched with `\x0`, but will be changed to `\n` which is POSIXLY supported) means that greediness will fail until the `\(.*\)` matches only the same text as the initial text after the separator, leaving the non-common prefix to be matched by the `.*` after parentheses..
